xc7/utils/prjxray_form_channels.py

```diff
@@ -929,6 +986,160 @@ def create_constant_tracks(conn):
+    return new_pkey
+
+def remap_tile_grid(conn, grid_map, tile_map, pip_map):
+    """
+    Remaps tiles present in the "phy_tile" table to the "tile" table.
+    :param conn:
+    :param grid_map:
+    :return:
+    """
+
+    c = conn.cursor()
+
+    # Get NULL tile pkey
+    null_pkey = c.execute("SELECT pkey FROM tile_type WHERE name = \"NULL\""
+                          ).fetchone()[0]
+
+    assert null_pkey is not None
+
+    # Fetch all physical tiles from the database
+    tile_data = c.execute(
+        "SELECT pkey, name, tile_type_pkey, grid_x, grid_y FROM phy_tile"
+    ).fetchall()
+
+    # Convert them to the new grid one-by-one
+    for tile in progressbar.progressbar(tile_data):
+        tile_type_pkey = tile[2]
+        phy_loc_x = tile[3]
+        phy_loc_y = tile[4]
+
+        # Get tile type string
+        tile_type = c.execute("SELECT name FROM tile_type WHERE pkey = (?)", (tile_type_pkey, )).fetchone()[0]
+
+        # Map location. Possibly one to many
+        vpr_locs = grid_map.get_vpr_loc((phy_loc_x, phy_loc_y))
+
+        # The tile is being split
+        if tile_type_pkey in tile_map.fwd_map.keys():
+
+            # Get base VPR location
+            base_vpr_loc = (
+                min([loc[0] for loc in vpr_locs]),
+                min([loc[1] for loc in vpr_locs])
+            )
+
+            # Fetch pip_name -> pip_pkey map from the database
+            pip_pkey_map = c.execute("SELECT name, pkey FROM pip_in_tile WHERE tile_type_pkey = (?)", (tile_type_pkey,)).fetchall()
+            pip_pkey_map = {m[0]: m[1] for m in pip_pkey_map}
+
+            # Insert tiles into the VPR grid
+            for vpr_tile_type_pkey, loc_ofs in tile_map.fwd_map[tile_type_pkey]:
+
+                # Compute location in the VPR grid
+                vpr_loc = (
+                    base_vpr_loc[0] + loc_ofs[0],
+                    base_vpr_loc[1] + loc_ofs[1],
+                )
+
+                # Get tile type as string
+                vpr_tile_type = c.execute("SELECT name FROM tile_type WHERE pkey = (?)", (vpr_tile_type_pkey, )).fetchone()[0]
+
+                # Generate new tile name
+                # FIXME: This will be the SLICE name. However it will not match
+                # FIXME: the Vivado slice name, coordinates will differ.
+                vpr_tile_name = "%s_X%dY%d" % (vpr_tile_type, vpr_loc[0], vpr_loc[1])
+
+                # Insert the tile into grid
+                vpr_tile_pkey = insert_vpr_tile(conn, vpr_tile_name, vpr_loc, vpr_tile_type_pkey, tile[0])
+
+                # Get site instance basing on the new tile type and site
+                # location offset. Here we assume that new tile type is
+                # equal to its former site type.
+                site_name, site_pkey = c.execute("SELECT name, pkey FROM site WHERE x_coord = (?) AND y_coord = (?) AND site_type_pkey = (SELECT pkey FROM site_type WHERE name = (?))", (loc_ofs[0], loc_ofs[1], vpr_tile_type)).fetchone()
```

Comment:
I don't understand why this is required.  Why doesn't the fwd_map provide the site_pkey?  This lookup seems error prone and not required.  Shouldn't this information already be encoded in a table?

Reply:
But the information itself is required. I guess it can be encoded in the fwd_map. I'll look into it.

Reply:
> But the information itself is required. I guess it can be encoded in the fwd_map. I'll look into it.

What I'm saying is this lookup seems fragile and redundant.  The tile splitter code already understood the tile<-> relationship, and should've have encoded in an unambiguous way.